ir/pipeline/block.rs

```diff
@@ -185,17 +197,23 @@ impl BlockContext {
     ) -> Result<(), Box<RepresentationError>> {
         debug_assert!(self.variable_declaration.contains_key(&var));
         self.add_referenced_variable(var);
-        let existing_scope = self.variable_declaration.get_mut(&var).unwrap();
-        if is_equal_or_parent_scope(&self.scope_parents, scope, *existing_scope) {
+        let recorded_scope = self.variable_declaration[&var];
+        if is_equal_or_parent_scope(&self.scope_parents, scope, recorded_scope) {
             // Parent defines same name: ok, reuse the variable
-            Ok(())
-        } else if is_child_scope(&self.scope_parents, scope, *existing_scope) {
+        } else if is_child_scope(&self.scope_parents, scope, recorded_scope) {
             // Child defines the same name: ok, reuse the variable, and change the declaration scope to the current one
-            *existing_scope = scope;
-            Ok(())
+            *self.variable_declaration.get_mut(&var).unwrap() = scope;
         } else {
-            Err(Box::new(RepresentationError::DisjointVariableReuse { name: var_name.to_string(), source_span }))
+            let ancestor = common_ancestor(&self.scope_parents, recorded_scope, scope);
+            if !self.is_visible_child(scope, ancestor) || !self.is_visible_child(recorded_scope, ancestor) {
+                return Err(Box::new(RepresentationError::DisjointVariableReuse {
+                    name: var_name.to_owned(),
+                    source_span,
+                }));
+            }
```

Comment:
Will this firing or not depend on which nested pattern is processed first? Do we just want to process negations at the end?
```
not { $x > 25; }; { $x isa age 5; } or {$x isa age 10; };
```

Reply:
Good point. I wonder if this error check can be removed altogether, surely the variable dependency check handles this already.

Reply:
Yeah, I wouldn't mind that. I think the scope being set to the common ancestor is still correct.